File: haskell/src/Surd/Algebraic/RootIsolation.hs
-- | Real root isolation for polynomials with rational coefficients.
--
-- Uses Sturm's theorem with bisection to separate and refine roots
-- of square-free polynomials.
module Surd.Algebraic.RootIsolation
  ( IsolatingInterval(..)
  , isolateRealRoots
  , refineRoot
  , rootInInterval
  , sturmCount
  ) where

import Surd.Polynomial.Univariate
import Surd.Internal.Interval (Interval(..))

-- | An isolating interval for a real root of a polynomial.
-- The polynomial has exactly one root in the open interval (lo, hi),
-- unless lo == hi in which case it's an exact rational root.
data IsolatingInterval = IsolatingInterval
  { iiPoly     :: !(Poly Rational)  -- ^ The (square-free) polynomial
  , iiInterval :: !Interval         -- ^ Isolating interval
  } deriving (Show)

-- | Isolate all real roots of a polynomial.
-- Returns isolating intervals ordered by increasing root value.
-- The polynomial is made square-free internally.
isolateRealRoots :: Poly Rational -> [IsolatingInterval]
isolateRealRoots p
  | degree p <= 0 = []
  | degree p == 1 =
      case unPoly p of
        [a, b] ->
          let r = -a / b
          in [IsolatingInterval (monicPoly p) (Interval r r)]
        _ -> []
  | otherwise =
      let p' = monicPoly p
          bound = rootBound p'
          iv = Interval (-bound) bound
      in map (IsolatingInterval p') $ isolateIn p' iv

-- | Refine an isolating interval until its width is less than epsilon.
refineRoot :: Rational -> IsolatingInterval -> IsolatingInterval
refineRoot eps (IsolatingInterval p iv) =
  IsolatingInterval p (refineInterval p eps iv)

-- | Check if a specific rational is a root within the interval.
rootInInterval :: IsolatingInterval -> Maybe Rational
rootInInterval (IsolatingInterval p (Interval l h))
  | l == h && evalPoly p l == 0 = Just l
  | otherwise = Nothing

-- Internal

-- | Cauchy bound: all roots satisfy |r| ≤ bound.
rootBound :: Poly Rational -> Rational
rootBound (Poly []) = 0
rootBound (Poly cs) =
  let lc = last cs
      ratios = map (\c -> abs (c / lc)) (init cs)
  in 1 + maximum (0 : ratios)

-- | Isolate roots within a given interval using bisection.
-- Precondition: p is square-free.
isolateIn :: Poly Rational -> Interval -> [Interval]
isolateIn p iv@(Interval l h)
  | sc == 0 = []
  | sc == 1 = [iv]
  | otherwise =
      let m = (l + h) / 2
      in if evalPoly p m == 0
         then Interval m m : isolateIn p (Interval l m) ++ isolateIn p (Interval m h)
         else isolateIn p (Interval l m) ++ isolateIn p (Interval m h)
  where
    sc = sturmCount p l h

-- | Count real roots in (a, b] using Sturm's theorem.
sturmCount :: Poly Rational -> Rational -> Rational -> Int
sturmCount p a b =
  let chain = sturmChain p
      va = signChangesAt chain a
      vb = signChangesAt chain b
  in va - vb

-- | Build the Sturm chain: p₀ = p, p₁ = p', pₖ₊₁ = -rem(pₖ₋₁, pₖ).
sturmChain :: Poly Rational -> [Poly Rational]
sturmChain p = p : p' : go p p'
  where
    p' = diffPoly p
    go _ b | degree b < 0 = []
    go a b =
      let (_, r) = divModPoly a b
          nr = scalePoly (-1) r
      in if degree nr < 0 then [] else nr : go b nr

-- | Count sign changes in the Sturm chain evaluated at x.
signChangesAt :: [Poly Rational] -> Rational -> Int
signChangesAt chain x =
  let vals = filter (/= 0) $ map (`evalPoly` x) chain
      pairs = zip vals (drop 1 vals)
  in length $ filter (\(a, b) -> signum a /= signum b) pairs

-- | Refine an interval by bisection until width < eps.
refineInterval :: Poly Rational -> Rational -> Interval -> Interval
refineInterval _ _ iv@(Interval l h) | l == h = iv
refineInterval p eps iv@(Interval l h)
  | h - l < eps = iv
  | otherwise =
      let m = (l + h) / 2
          fm = evalPoly p m
      in if fm == 0
         then Interval m m
         else let fl = evalPoly p l
              in if signum fl /= signum fm
                 then refineInterval p eps (Interval l m)
                 else refineInterval p eps (Interval m h)

File: mercury/root_isolation.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: root_isolation.m
% Main author: nyc
%
% Real root isolation for polynomials with rational coefficients.
%
% Uses Sturm's theorem with bisection to separate and refine roots
% of square-free polynomials.
%
%---------------------------------------------------------------------------%

:- module root_isolation.
:- interface.

:- import_module interval.
:- import_module list.
:- import_module maybe.
:- import_module poly.
:- import_module rational.

%---------------------------------------------------------------------------%

    % An isolating interval for a real root of a polynomial.
    % The polynomial has exactly one root in the open interval (lo, hi),
    % unless lo = hi in which case it's an exact rational root.
    %
:- type isolating_interval
    --->    isolating_interval(
                ii_poly     :: poly(rational),
                ii_interval :: interval
            ).

    % Isolate all real roots of a polynomial.
    % Returns isolating intervals ordered by increasing root value.
    %
:- func isolate_real_roots(poly(rational)) = list(isolating_interval).

    % Refine an isolating interval until its width is less than epsilon.
    %
:- func refine_root(rational, isolating_interval) = isolating_interval.

    % Check if an exact rational root exists within the interval.
    %
:- func root_in_interval(isolating_interval) = maybe(rational).

    % Count real roots of p in the half-open interval (a, b]
    % using Sturm's theorem.
    %
:- func sturm_count(poly(rational), rational, rational) = int.

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- use_module int.
:- import_module integer.
:- import_module root_bound.

%---------------------------------------------------------------------------%
% Root isolation
%---------------------------------------------------------------------------%

isolate_real_roots(P) = Result :-
    D = degree(P),
    ( if int.'=<'(D, 0) then
        Result = []
    else if D = 1 then
        Cs = coeffs(P),
        ( if Cs = [A, B] then
            R = rational.'-'(rational.zero,
                rational.'/'(A, B)),
            Result = [isolating_interval(monic_poly(P),
                interval.interval(R, R))]
        else
            Result = []
        )
    else
        MP = monic_poly(P),
        Bound = root_bound.root_bound(MP),
        IV = interval.interval(rational.'-'(rational.zero, Bound), Bound),
        Result = list.map(
            ( func(I) = isolating_interval(MP, I) ),
            isolate_in(MP, IV))
    ).

%---------------------------------------------------------------------------%
% Refinement
%---------------------------------------------------------------------------%

refine_root(Eps, isolating_interval(P, IV)) =
    isolating_interval(P, refine_interval(P, Eps, IV)).

root_in_interval(isolating_interval(P, interval(L, H))) = Result :-
    ( if L = H, poly.eval(P, L) = rational.zero then
        Result = yes(L)
    else
        Result = no
    ).

%---------------------------------------------------------------------------%
% Internal: isolate within an interval
%---------------------------------------------------------------------------%

:- func isolate_in(poly(rational), interval) = list(interval).

isolate_in(P, IV) = Result :-
    IV = interval(L, H),
    SC = sturm_count(P, L, H),
    ( if SC = 0 then
        Result = []
    else if SC = 1 then
        Result = [IV]
    else
        Two = rational(2, 1),
        M = rational.'/'(rational.'+'(L, H), Two),
        FM = poly.eval(P, M),
        ( if FM = rational.zero then
            Result = [interval(M, M) |
                list.append(isolate_in(P, interval(L, M)),
                    isolate_in(P, interval(M, H)))]
        else
            Result = list.append(
                isolate_in(P, interval(L, M)),
                isolate_in(P, interval(M, H)))
        )
    ).

%---------------------------------------------------------------------------%
% Sturm chain and counting
%---------------------------------------------------------------------------%

sturm_count(P, A, B) = Result :-
    Chain = sturm_chain(P),
    VA = sign_changes_at(Chain, A),
    VB = sign_changes_at(Chain, B),
    Result = int.'-'(VA, VB).

:- func sturm_chain(poly(rational)) = list(poly(rational)).

sturm_chain(P) = [P, DP | build_chain(P, DP)] :-
    DP = poly.diff(P).

:- func build_chain(poly(rational), poly(rational)) = list(poly(rational)).

build_chain(A, B) = Result :-
    ( if int.'<'(degree(B), 0) then
        Result = []
    else
        poly.div_mod(A, B, _, R),
        NR = poly.scale(rational.'-'(rational.zero, rational.one), R),
        ( if int.'<'(degree(NR), 0) then
            Result = []
        else
            Result = [NR | build_chain(B, NR)]
        )
    ).

:- func sign_changes_at(list(poly(rational)), rational) = int.

sign_changes_at(Chain, X) = Result :-
    Vals0 = list.map(( func(P) = poly.eval(P, X) ), Chain),
    Vals = list.filter(( pred(V::in) is semidet :- V \= rational.zero ), Vals0),
    Result = count_sign_changes(Vals).

:- func count_sign_changes(list(rational)) = int.

count_sign_changes(Vals) = Result :-
    ( if Vals = [] then
        Result = 0
    else if Vals = [_] then
        Result = 0
    else if Vals = [A, B | Rest] then
        SA = rat_signum(A),
        SB = rat_signum(B),
        ( if SA \= SB then
            Result = int.'+'(1, count_sign_changes([B | Rest]))
        else
            Result = count_sign_changes([B | Rest])
        )
    else
        Result = 0
    ).

:- func rat_signum(rational) = int.

rat_signum(R) = Result :-
    ( if rational.'>'(R, rational.zero) then
        Result = 1
    else if rational.'<'(R, rational.zero) then
        Result = -1
    else
        Result = 0
    ).

%---------------------------------------------------------------------------%
% Make monic
%---------------------------------------------------------------------------%

:- func monic_poly(poly(rational)) = poly(rational).

monic_poly(P) = Result :-
    Cs = coeffs(P),
    ( if Cs = [] then
        Result = P
    else
        LC = list.det_last(Cs),
        ( if LC = rational.zero then
            Result = P
        else
            LCInv = rational.'/'(rational.one, LC),
            Result = poly.scale(LCInv, P)
        )
    ).

%---------------------------------------------------------------------------%
% Interval refinement
%---------------------------------------------------------------------------%

:- func refine_interval(poly(rational), rational, interval) = interval.

refine_interval(P, Eps, IV) = Result :-
    IV = interval(L, H),
    ( if L = H then
        Result = IV
    else if rational.'<'(rational.'-'(H, L), Eps) then
        Result = IV
    else
        Two = rational(2, 1),
        M = rational.'/'(rational.'+'(L, H), Two),
        FM = poly.eval(P, M),
        ( if FM = rational.zero then
            Result = interval(M, M)
        else
            FL = poly.eval(P, L),
            ( if rat_signum(FL) \= rat_signum(FM) then
                Result = refine_interval(P, Eps, interval(L, M))
            else
                Result = refine_interval(P, Eps, interval(M, H))
            )
        )
    ).

%---------------------------------------------------------------------------%
:- end_module root_isolation.
%---------------------------------------------------------------------------%
